File: go-backend/internal/models/song.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Song struct {
	ID          primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Title       string             `json:"title" bson:"title"`
	Artist      []string           `json:"artist" bson:"artist"`
	Album       string             `json:"album" bson:"album"`
	Genre       string             `json:"genre" bson:"genre"`
	Duration    int                `json:"duration" bson:"duration"` // in seconds
	FilePath    string             `json:"filePath" bson:"filePath"`
	CoverImage  string             `json:"coverImage" bson:"coverImage"`
	ReleaseDate time.Time          `json:"releaseDate" bson:"releaseDate"`
	CreatedAt   time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt   time.Time          `json:"updatedAt" bson:"updatedAt"`
}

type SongResponse struct {
	ID          string    `json:"id"`
	Title       string    `json:"title"`
	Artist      []string  `json:"artist"`
	Album       string    `json:"album"`
	Genre       string    `json:"genre"`
	Duration    int       `json:"duration"`
	FilePath    string    `json:"filePath"`
	CoverImage  string    `json:"coverImage"`
	ReleaseDate time.Time `json:"releaseDate"`
}

func (s *Song) ToResponse() *SongResponse {
	return &SongResponse{
		ID:          s.ID.Hex(),
		Title:       s.Title,
		Artist:      s.Artist,
		Album:       s.Album,
		Genre:       s.Genre,
		Duration:    s.Duration,
		FilePath:    s.FilePath,
		CoverImage:  s.CoverImage,
		ReleaseDate: s.ReleaseDate,
	}
}

File: node-backend/src/models/Song.js
const mongoose = require('mongoose');

const songSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Song title is required'],
    trim: true,
    maxlength: [200, 'Title cannot exceed 200 characters']
  },
  artist: [{
    type: String,
    required: [true, 'At least one artist is required'],
    trim: true,
    maxlength: [100, 'Artist name cannot exceed 100 characters']
  }],
  album: {
    type: String,
    trim: true,
    maxlength: [200, 'Album name cannot exceed 200 characters']
  },
  genre: {
    type: String,
    trim: true,
    maxlength: [50, 'Genre cannot exceed 50 characters']
  },
  duration: {
    type: Number,
    required: [true, 'Song duration is required'],
    min: [1, 'Duration must be at least 1 second']
  },
  filePath: {
    type: String,
    required: [true, 'File path is required']
  },
  coverImage: {
    type: String,
    default: ''
  },
  releaseDate: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true,
  toJSON: {
    transform: function(doc, ret) {
      ret.id = ret._id;
      delete ret._id;
      delete ret.__v;
      return ret;
    }
  },
  toObject: {
    transform: function(doc, ret) {
      ret.id = ret._id;
      delete ret._id;
      delete ret.__v;
      return ret;
    }
  }
});

// Get song response format
songSchema.methods.toResponse = function() {
  const song = this.toObject();
  return {
    id: song.id,
    title: song.title,
    artist: song.artist,
    album: song.album,
    genre: song.genre,
    duration: song.duration,
    filePath: song.filePath,
    coverImage: song.coverImage,
    releaseDate: song.releaseDate
  };
};

// Indexes for better search performance
songSchema.index({ title: 'text', artist: 'text', album: 'text' });
songSchema.index({ genre: 1 });
songSchema.index({ artist: 1 });
songSchema.index({ createdAt: -1 });

module.exports = mongoose.model('Song', songSchema);